src/main/java/org/apache/mesos/scheduler/SchedulerDriverFactory.java

```diff
@@ -0,0 +1,84 @@
+import org.apache.mesos.Scheduler;
+import org.apache.mesos.SchedulerDriver;
+import org.apache.mesos.Protos.Credential;
+import org.apache.mesos.Protos.FrameworkInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.google.protobuf.ByteString;
+
+/**
+ * Factory class for creating {@link MesosSchedulerDriver}s.
+ */
+public class SchedulerDriverFactory {
+
+  private static final Logger LOGGER =
+    LoggerFactory.getLogger(SchedulerDriverFactory.class);
+
+  /**
+   * Creates and returns a new {@link SchedulerDriver} without a credential secret.
+   *
+   * @param scheduler The Framework {@link Scheduler} implementation which should receive callbacks
+   * from the {@link SchedulerDriver}
+   * @param frameworkInfo The {@link FrameworkInfo} which describes the framework implementation.
+   * The 'principal' field MUST be populated and non-empty
+   * @param masterZkUrl The URL of the currently active Mesos Master, of the form "zk://host/mesos"
+   * @return A {@link SchedulerDriver} configured with the provided info
+   * @throws IllegalArgumentException if {@link FrameworkInfo}.principal is unset or empty
+   */
+  public SchedulerDriver create(
+      final Scheduler scheduler, final FrameworkInfo frameworkInfo, final String masterZkUrl) {
+    return create(scheduler, frameworkInfo, masterZkUrl, null /* credentialSecret */);
+  }
+
+  /**
+   * Creates and returns a new {@link SchedulerDriver} with the provided credential secret.
+   *
+   * @param scheduler The Framework {@link Scheduler} implementation which should receive callbacks
+   * from the {@link SchedulerDriver}
+   * @param frameworkInfo The {@link FrameworkInfo} which describes the framework implementation.
+   * The 'principal' field MUST be populated and non-empty
+   * @param masterZkUrl The URL of the currently active Mesos Master, of the form "zk://host/mesos"
+   * @param credentialSecret The secret to be included in the framework
+   *     {@link org.apache.mesos.Protos.Credential}, ignored if {@code null}/empty
+   * @return A {@link SchedulerDriver} configured with the provided info
+   * @throws IllegalArgumentException if {@link FrameworkInfo}.principal is unset or empty
+   */
+  public SchedulerDriver create(
+      final Scheduler scheduler,
+      final FrameworkInfo frameworkInfo,
+      final String masterZkUrl,
+      final byte[] credentialSecret) {
+    LOGGER.info("Creating MesosSchedulerDriver for "
+        + "scheduler = {}, frameworkInfo = {}, masterZkUrl = {}, credentialSecret = {}",
+        scheduler, frameworkInfo, masterZkUrl, credentialSecret);
+
+    if (!frameworkInfo.hasPrincipal() || StringUtils.isEmpty(frameworkInfo.getPrincipal())) {
+      throw new IllegalArgumentException(
+          "Unable to create MesosSchedulerDriver, FrameworkInfo lacks a principal: "
+              + frameworkInfo.toString());
+    }
+
+    Credential.Builder credentialBuilder = Credential.newBuilder()
```

Comment:
However, always generating credentials shouldn't be a requirement.  In an unauthenticated environment, I'm not sure if providing unneeded credentials will result in everything being ok, or in things not working.  Better to not generate credentials when we expect them not to be used.